File: examples/python/apps/migrating_merlin/sweep_actor.py
# SPDX-License-Identifier: LGPL-2.1-or-later
# Copyright (C) 2025 PlexSpaces Contributors
#
# Merlin → PlexSpaces: Scientific simulation manager (parameter sweep).
#
# Worker pool (elastic size): elastic pool API (checkout/checkin) + tuple space (work queue).
# - Coordinator (run): scatter tasks → checkout workers from pool, send work_available → gather → checkin.
#   If pool is not configured, falls back to process group broadcast.
# - Workers (work_available): take tasks from tuple space, run simulation, write results.
# Convention: sweep:coord-1 = coordinator; sweep:worker-0, worker-1, ... = pool workers.

from plexspaces import workflow_actor, state, handler, host

SIMULATION_MS = 22
TUPLE_PREFIX = "merlin"
WORKER_GROUP = "merlin-workers"
POOL_NAME = "merlin-workers"
MAX_CHECKOUT_WORKERS = 10
GATHER_POLL_MAX = 200
CHECKOUT_TIMEOUT_MS = 5000


@workflow_actor(facets=["virtual_actor", "durability"])
class SweepActor:
    """
    Parameter sweep: run() = coordinator (scatter → checkout workers from pool, send work → gather → checkin);
    work_available = worker (take params from tuple space, run simulation, write results).
    Uses elastic pool API when available; falls back to process group broadcast otherwise.
    """

    sweep_id: str = state(default="")
    num_params: int = state(default=0)
    num_completed: int = state(default=0)
    status: str = state(default="idle")
    total_compute_ms: float = state(default=0.0)
    total_coord_ms: float = state(default=0.0)
    created_at_ms: float = state(default=0.0)
    updated_at_ms: float = state(default=0.0)
    cancel_requested: bool = state(default=False)
    worker_joined: bool = state(default=False)

    def run(self, payload: dict) -> dict:
        """Coordinator: scatter tasks, checkout workers from pool and send work_available (or broadcast), gather, checkin."""
        t0 = host.now_ms()
        if self.created_at_ms == 0:
            self.created_at_ms = float(t0)
        self.sweep_id = str(
            payload.get("sweep_id") or payload.get("job_id") or self.sweep_id
        )
        self.updated_at_ms = float(host.now_ms())

        if self.cancel_requested:
            self.status = "cancelled"
            return self._finish(t0, 0.0, "cancelled")

        if self.status == "completed":
            return self._finish(t0, 0.0, "completed")

        num_params = int(payload.get("num_params", 10))
        if num_params <= 0:
            return self._finish(t0, 0.0, "no_params")

        self.num_params = num_params
        self.num_completed = 0
        self.status = "scattering"
        compute_ms = 0.0

        # Tuple space: scatter parameter-sweep tasks (list-in/list-out)
        for i in range(num_params):
            param_val = payload.get("param_base", 0) + i
            out = host.ts.write(
                [TUPLE_PREFIX, self.sweep_id, "task", f"p{i}", {"param": param_val}]
            )
            if out and out.startswith("ERROR"):
                host.log("warn", f"ts write failed: {out}")
            compute_ms += 2
        self.updated_at_ms = host.now_ms()
        self.status = "running"

        # Worker pool: use elastic pool API (checkout → send work → gather → checkin); fallback to process group broadcast
        checkout_handles: list = []
        try:
            for _ in range(min(MAX_CHECKOUT_WORKERS, num_params)):
                handle = host.pool_checkout(POOL_NAME, CHECKOUT_TIMEOUT_MS)
                if handle is None:
                    break
                checkout_handles.append(handle)
                out = host.send(
                    handle["actor_id"],
                    "work_available",
                    {"sweep_id": self.sweep_id, "num_params": num_params},
                )
                if out and out.startswith("ERROR"):
                    host.log("warn", f"send to worker failed: {out}")
            if not checkout_handles:
                host.process_groups.broadcast(
                    WORKER_GROUP,
                    "work_available",
                    {"sweep_id": self.sweep_id, "num_params": num_params},
                )
        except Exception as e:
            host.log("warn", f"pool/broadcast failed: {e}")
            try:
                host.process_groups.broadcast(
                    WORKER_GROUP,
                    "work_available",
                    {"sweep_id": self.sweep_id, "num_params": num_params},
                )
            except Exception as e2:
                host.log("warn", f"pg_broadcast failed: {e2}")
        self.total_coord_ms += float(host.now_ms() - t0)

        # Gather results from tuple space
        pattern = [TUPLE_PREFIX, self.sweep_id, "result", None, None]
        for _ in range(GATHER_POLL_MAX):
            if self.cancel_requested:
                self.status = "cancelled"
                return self._finish(t0, compute_ms, "cancelled")
            results = host.ts.read_all(pattern)
            self.num_completed = len(results) if results else 0
            if self.num_completed >= num_params:
                break
            self.updated_at_ms = host.now_ms()

        # Checkin all workers that were checked out from the pool
        for handle in checkout_handles:
            try:
                host.pool_checkin(
                    POOL_NAME,
                    handle["actor_id"],
                    handle["checkout_id"],
                    True,
                )
            except Exception as e:
                host.log("warn", f"pool_checkin failed: {e}")

        self.status = "completed"
        self.updated_at_ms = host.now_ms()
        return self._finish(t0, compute_ms, "completed")

    def _finish(self, t0: float, compute_ms: float, status: str) -> dict:
        elapsed = float(host.now_ms() - t0)
        if elapsed < compute_ms:
            elapsed = compute_ms
        coord_ms = max(0.0, elapsed - compute_ms)
        self.total_compute_ms += compute_ms
        self.total_coord_ms += coord_ms
        # Approximate data size: each task tuple ~100 bytes, result ~80 bytes
        approx_bytes = (self.num_params * 100) + (self.num_completed * 80)
        return {
            "status": status,
            "sweep_id": self.sweep_id,
            "num_params": self.num_params,
            "num_completed": self.num_completed,
            "data_size_bytes": approx_bytes,
            "total_compute_ms": self.total_compute_ms,
            "total_coord_ms": self.total_coord_ms,
        }

    def signal(self, name: str, data: dict) -> None:
        if name == "cancel":
            self.cancel_requested = True
            self.updated_at_ms = host.now_ms()

    def query(self, name: str, params: dict) -> dict:
        if name == "status":
            return {
                "sweep_id": self.sweep_id,
                "status": self.status,
                "num_params": self.num_params,
                "num_completed": self.num_completed,
                "cancel_requested": self.cancel_requested,
                "total_compute_ms": self.total_compute_ms,
                "total_coord_ms": self.total_coord_ms,
                "created_at_ms": self.created_at_ms,
                "updated_at_ms": self.updated_at_ms,
            }
        return {"error": "unknown_query", "name": name}

    @handler("work_available")
    def on_work_available(
        self,
        sweep_id: str = "",
        num_params: int = 0,
        **kwargs,
    ) -> dict:
        """Worker: take tasks from tuple space, run simulation, write results. Joins process group for broadcast fallback."""
        if not self.worker_joined:
            try:
                host.process_groups.join(WORKER_GROUP)
                self.worker_joined = True
                host.log("info", f"Joined worker pool {WORKER_GROUP}")
            except Exception as e:
                host.log("warn", f"pg_join failed: {e}")
        if not sweep_id or num_params <= 0:
            return {"ok": True, "message": "warmup"}
        t0 = host.now_ms()
        pattern = [TUPLE_PREFIX, sweep_id, "task", None, None]
        processed = 0
        compute_ms = 0.0
        while True:
            taken = host.ts.take(pattern)
            if taken is None:
                break
            param_id = taken[3] if len(taken) > 3 else ""
            compute_ms += SIMULATION_MS
            host.ts.write(
                [
                    TUPLE_PREFIX,
                    sweep_id,
                    "result",
                    param_id,
                    {"ok": True, "ms": SIMULATION_MS},
                ]
            )
            processed += 1
        elapsed = float(host.now_ms() - t0)
        self.total_compute_ms += compute_ms
        self.total_coord_ms += max(0.0, elapsed - compute_ms)
        return {"ok": True, "processed": processed, "sweep_id": sweep_id}

File: examples/typescript/apps/migrating_merlin/sweep_actor.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2025 PlexSpaces Contributors
//
// Merlin → PlexSpaces: Parameter sweep (TypeScript WASM).
//
// Worker pool (elastic size): elastic pool API (poolCheckout/poolCheckin) + tuple space (work queue).
// Coordinator (run): scatter → checkout workers from pool, send work_available → gather → checkin.
// If pool is not configured, falls back to process group broadcast.
// Workers (onWork_available): take tasks from tuple space, run simulation, write results.

import { WorkflowActor, host } from "@plexspaces/sdk";

const SIMULATION_MS = 22;
const TUPLE_PREFIX = "merlin";
const WORKER_GROUP = "merlin-workers";
const POOL_NAME = "merlin-workers";
const MAX_CHECKOUT_WORKERS = 10;
const GATHER_POLL_MAX = 200;
const CHECKOUT_TIMEOUT_MS = 5000;

interface SweepState {
  sweep_id: string;
  num_params: number;
  num_completed: number;
  status: string;
  total_compute_ms: number;
  total_coord_ms: number;
  created_at_ms: number;
  updated_at_ms: number;
  cancel_requested: boolean;
  worker_joined: boolean;
}

export class SweepActor extends WorkflowActor<SweepState> {
  getDefaultState(): SweepState {
    return {
      sweep_id: "",
      num_params: 0,
      num_completed: 0,
      status: "idle",
      total_compute_ms: 0,
      total_coord_ms: 0,
      created_at_ms: 0,
      updated_at_ms: 0,
      cancel_requested: false,
      worker_joined: false,
    };
  }

  run(payload: Record<string, unknown>): Record<string, unknown> {
    const t0 = host.nowMs();
    if (this.state.created_at_ms === 0) this.state.created_at_ms = t0;
    this.state.sweep_id = String(payload.sweep_id ?? payload.job_id ?? this.state.sweep_id);
    this.state.updated_at_ms = host.nowMs();

    if (this.state.cancel_requested) {
      this.state.status = "cancelled";
      return this.finish(t0, 0, "cancelled");
    }
    if (this.state.status === "completed") return this.finish(t0, 0, "completed");

    const numParams = Number(payload.num_params ?? 10) || 0;
    if (numParams <= 0) return this.finish(t0, 0, "no_params");

    this.state.num_params = numParams;
    this.state.num_completed = 0;
    this.state.status = "scattering";
    let computeMs = 0;
    const paramBase = Number(payload.param_base ?? 0) || 0;

    for (let i = 0; i < numParams; i++) {
      const paramVal = paramBase + i;
      const out = host.ts.write([
        TUPLE_PREFIX,
        this.state.sweep_id,
        "task",
        `p${i}`,
        { param: paramVal },
      ]);
      if (out && out.startsWith("ERROR")) host.log("warn", `ts write failed: ${out}`);
      computeMs += 2;
    }
    this.state.updated_at_ms = host.nowMs();
    this.state.status = "running";

    const checkoutHandles: { actor_id: string; pool_name: string; checkout_id: string }[] = [];
    try {
      for (let i = 0; i < Math.min(MAX_CHECKOUT_WORKERS, numParams); i++) {
        const handle = host.poolCheckout(POOL_NAME, CHECKOUT_TIMEOUT_MS);
        if (handle == null) break;
        checkoutHandles.push(handle);
        const out = host.send(handle.actor_id, "work_available", {
          sweep_id: this.state.sweep_id,
          num_params: numParams,
        });
        if (out && out.startsWith("ERROR")) host.log("warn", `send to worker failed: ${out}`);
      }
      if (checkoutHandles.length === 0) {
        host.processGroups.broadcast(WORKER_GROUP, "work_available", {
          sweep_id: this.state.sweep_id,
          num_params: numParams,
        });
      }
    } catch (e) {
      host.log("warn", `pool/broadcast failed: ${e}`);
      try {
        host.processGroups.broadcast(WORKER_GROUP, "work_available", {
          sweep_id: this.state.sweep_id,
          num_params: numParams,
        });
      } catch (e2) {
        host.log("warn", `pg_broadcast failed: ${e2}`);
      }
    }
    this.state.total_coord_ms += host.nowMs() - t0;

    const pattern: unknown[] = [TUPLE_PREFIX, this.state.sweep_id, "result", null, null];
    for (let i = 0; i < GATHER_POLL_MAX; i++) {
      if (this.state.cancel_requested) {
        this.state.status = "cancelled";
        return this.finish(t0, computeMs, "cancelled");
      }
      const results = host.ts.readAll(pattern);
      this.state.num_completed = Array.isArray(results) ? results.length : 0;
      if (this.state.num_completed >= numParams) break;
      this.state.updated_at_ms = host.nowMs();
    }

    for (const h of checkoutHandles) {
      try {
        host.poolCheckin(POOL_NAME, h.actor_id, h.checkout_id, true);
      } catch (e) {
        host.log("warn", `pool_checkin failed: ${e}`);
      }
    }

    this.state.status = "completed";
    this.state.updated_at_ms = host.nowMs();
    return this.finish(t0, computeMs, "completed");
  }

  private finish(
    t0: number,
    computeMs: number,
    status: string
  ): Record<string, unknown> {
    const elapsed = host.nowMs() - t0;
    const coordMs = Math.max(0, (elapsed > computeMs ? elapsed : computeMs) - computeMs);
    this.state.total_compute_ms += computeMs;
    this.state.total_coord_ms += coordMs;
    const approxBytes = this.state.num_params * 100 + this.state.num_completed * 80;
    return {
      status,
      sweep_id: this.state.sweep_id,
      num_params: this.state.num_params,
      num_completed: this.state.num_completed,
      data_size_bytes: approxBytes,
      total_compute_ms: this.state.total_compute_ms,
      total_coord_ms: this.state.total_coord_ms,
    };
  }

  signal(name: string, _data: Record<string, unknown>): void {
    if (name === "cancel") {
      this.state.cancel_requested = true;
      this.state.updated_at_ms = host.nowMs();
    }
  }

  query(name: string, _params: Record<string, unknown>): Record<string, unknown> {
    if (name === "status") {
      return {
        sweep_id: this.state.sweep_id,
        status: this.state.status,
        num_params: this.state.num_params,
        num_completed: this.state.num_completed,
        cancel_requested: this.state.cancel_requested,
        total_compute_ms: this.state.total_compute_ms,
        total_coord_ms: this.state.total_coord_ms,
        created_at_ms: this.state.created_at_ms,
        updated_at_ms: this.state.updated_at_ms,
      };
    }
    return { error: "unknown_query", name };
  }

  /** Worker: take tasks from tuple space, run simulation, write results. Joins process group for broadcast fallback. */
  onWork_available(payload: Record<string, unknown>): Record<string, unknown> {
    if (!this.state.worker_joined) {
      try {
        host.processGroups.join(WORKER_GROUP);
        this.state.worker_joined = true;
        host.log("info", `Joined worker pool ${WORKER_GROUP}`);
      } catch (e) {
        host.log("warn", `pg_join failed: ${e}`);
      }
    }
    const sweepId = String(payload.sweep_id ?? "");
    const numParams = Number(payload.num_params ?? 0) || 0;
    if (!sweepId || numParams <= 0) return { ok: true, message: "warmup" };

    const t0 = host.nowMs();
    const pattern: unknown[] = [TUPLE_PREFIX, sweepId, "task", null, null];
    let processed = 0;
    let computeMs = 0;
    while (true) {
      const taken = host.ts.take(pattern);
      if (taken == null) break;
      const paramId = Array.isArray(taken) && taken.length > 3 ? taken[3] : "";
      computeMs += SIMULATION_MS;
      host.ts.write([
        TUPLE_PREFIX,
        sweepId,
        "result",
        paramId,
        { ok: true, ms: SIMULATION_MS },
      ]);
      processed++;
    }
    const elapsed = host.nowMs() - t0;
    this.state.total_compute_ms += computeMs;
    this.state.total_coord_ms += Math.max(0, elapsed - computeMs);
    return { ok: true, processed, sweep_id: sweepId };
  }
}

const actorInstance = new SweepActor();
export const actor = {
  init: (configJson: string) => actorInstance.init(configJson),
  handle: (from: string, msgType: string, payloadJson: string) =>
    actorInstance.handle(from, msgType, payloadJson),
  getState: () => actorInstance.getState(),
  setState: (stateJson: string) => actorInstance.setState(stateJson),
};
